### DISCUSSION TITLE
Unable to customize branch name for dependency which has version as a Maven property

### DISCUSSION THREAD

**Discussion by noelmccrory** [renovatebot/renovate#8002] [Request Help] [answered]
In my Maven pom.xml, I have:

```
<properties>
    <testcontainers.version>1.15.0</testcontainers.version>
</properties>
```

In my renovate config, I have:

```
branchTopic: "{{{depNameSanitized}}}-{{{toVersion}}}"
```

However, when a branch is created for a test-containers update, it has a name of `renovate/testcontainers.version`.

If the dependency version is declared in-line, e.g.

```
    <dependency>
      <groupId>com.github.ben-manes.caffeine</groupId>
      <artifactId>caffeine</artifactId>
      <version>2.8.7</version>
    </dependency>
```

Then it is created with the expected name, e.g. `renovate/com.github.ben-manes.caffeine-caffeine-2.8.8`

**Comment by rarkins** [comment, marked as answer]
1. If you are using properties, you should update all dependencies that share it together, because.. they will be upgraded together whether you like it or not and you'd otherwise just end up with multiple PRs all making the same change.

2. If you only have one dependency per property, reconsider if you can avoid it

In other words: if you have multiple dependencies sharing the same version then it's a mistake to divide them into separate PRs which update identical things, and if you have only one dependency per version then (a) why do that? (b) is this just a cosmetic thing or is there a functional reason you need to change it?

**Reply by noelmccrory** [reply]
So for the testcontainers example above, there are multiple dependencies that use the same version property:

```
    <dependency>
      <groupId>org.testcontainers</groupId>
      <artifactId>postgresql</artifactId>
      <version>${testcontainers.version}</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.testcontainers</groupId>
      <artifactId>localstack</artifactId>
      <version>${testcontainers.version}</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.testcontainers</groupId>
      <artifactId>spock</artifactId>
      <version>${testcontainers.version}</version>
      <scope>test</scope>
    </dependency>
```

I want a single PR to be created when an update occurs for testcontainers and that is what's currently happening.

The problem that I'm having is with customizing the name of the branch that it creates. I would like to be able to customize the branch name to be e.g. `renovate/testcontainers.version-1.15.1` , as per the caffeine example above, not just `renovate/testcontainers.version`

The reason for this is that we are using Bamboo and branch plans gets automatically disabled after the branch has been merged/deleted. So when renovate runs a number of days later and recreates the branch in BitBucket, Bamboo does not automatically re-activate the branch plan. So my solution was to try and make each branch name unique by including the full version number in it.

**Reply by viceice** [reply]
I think you need a package rule, with set's branch topic and resets group name. Group name will override the branch name otherwise.

**Reply by viceice** [reply]
It would be very helpful to have a minimal reproduction repo.

**Reply by rarkins** [reply]
This is the result of a little quick in how Maven handles property versions. It was essential that we group them together for reasons mentioned above so it actually sets a `groupName` as part of the extract process: https://github.com/renovatebot/renovate/blob/f1bf05756ac298fdd077c086625c8f64d1e62296/lib/manager/maven/extract.ts#L111-L114

`groupName` takes a higher priority than most things and so yes it's not ideal for being able to change.

I would be interested in trying to add the version to the branch name automatically though.

@noelmccrory does your Bamboo process mean that you can *never* reuse a branch name? If so then it's probably going to cause you some challenges elsewhere too.

**Reply by noelmccrory** [reply]
I was able to work around this by removing the Maven version property and adding a package rule.

Before:

pom.xml
```
<properties>
  <springfox.version>3.0.0</springfox.version>
</properties>
 
<dependencies>
  <dependency>
    <groupId>io.springfox</groupId>
    <artifactId>springfox-boot-starter</artifactId>
    <version>${springfox.version}</version>
  </dependency>
  <dependency>
    <groupId>io.springfox</groupId>
    <artifactId>springfox-swagger-ui</artifactId>
    <version>${springfox.version}</version>
  </dependency>
</dependencies>
```

After:

pom.xml
```
<properties>
</properties>
 
<dependencies>
  <dependency>
    <groupId>io.springfox</groupId>
    <artifactId>springfox-boot-starter</artifactId>
    <version>3.0.0</version>
  </dependency>
  <dependency>
    <groupId>io.springfox</groupId>
    <artifactId>springfox-swagger-ui</artifactId>
    <version>3.0.0</version>
  </dependency>
</dependencies>
```

renovate.json
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "packageRules": [
    {
      packageNames: ["io.springfox:springfox-swagger-ui", "io.springfox:springfox-boot-starter"],
      branchTopic: "springfox-{{{toVersion}}}"
    }
  ]
}
```
